persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala

```diff
@@ -16,110 +16,45 @@
-       throw new IllegalArgumentException("Exhausted retries.", attempts.failed.get)
-    }
+  def save[T, R](entity: T, attemptSave: T => Future[R]): Future[R] = {
+    val attempts: Try[Future[R]] = saveWithRetry(entity, attemptSave, TotalRetriesOnConflictingKey)
+    attempts.getOrElse(throw new IllegalArgumentException("Exhausted retries.", attempts.failed.get))
   }
 
   @tailrec
-  private def saveWithRetry(dataLineage: DataLineage, retries: Int): Try[Unit] = {
+  private def saveWithRetry[T, R](entity: T, attemptSave: T => Future[R], retries: Int): Try[Future[R]] = {
     val left = retries - 1
-    Try(attemptSave(dataLineage))
-      match {
-        case s: Success[Unit] => s
-        case Failure(RetryableException(e)) =>
-            if (retries == 0) {
-              Failure(e)
-            } else {
-              log.warn(s"Ignoring ${e.getClass.getSimpleName} and $left left. Exception message: ${e.getMessage}.")
-              saveWithRetry(dataLineage, left)
-            }
-        case Failure(e) =>
-          throw new IllegalArgumentException(s"Unexpected exception aborting remaining $left retries.", e)
-      }
-  }
-
-  private def attemptSave(dataLineage: DataLineage): Unit = {
-    val uris = referencedUris(dataLineage)
-    val uriExistingKey = queryExistingToKey(uris)
-    val uriToNewKey = generateNewKeys(uris, uriExistingKey)
-    val uriToKey = uriExistingKey ++ uriToNewKey
-    val params = DataLineageTransactionParams
-      .create(dataLineage, uriToNewKey, uriToKey)
-    val options = new TransactionOptions()
-      .params(params) // Serialized hash map with json string values.
-      .writeCollections(params.fields: _*)
-      .allowImplicit(false)
-    val action: String = s"""
-        |function (params) {
-        |  var db = require('internal').db;
-        |  ${params.saveCollectionsJs}
-        |}
-        |""".stripMargin
-    db.transaction(action, classOf[Void], options)
-  }
-
-  private def referencedUris(dataLineage: DataLineage) = {
-    dataLineage.operations
-      .flatMap(op => op match {
-        case r: splinemodel.op.Read => r.sources.map(s => s.path)
-        case w: splinemodel.op.Write => Some(w.path)
-        case _ => None })
-      .distinct
-  }
-
-  private def queryExistingToKey(uris: Seq[String]): Map[String, String] = {
-    val urisList = uris
-      .map(uri => "\"" + uri + "\"")
-      .mkString(", ")
-    val query = s"for ds in dataSource filter ds.uri in [$urisList] return ds"
-    val result = db.query(query, classOf[VPackSlice])
-    result
-      .asInstanceOf[java.util.Iterator[VPackSlice]]
-      .asScala
-      .map(vpack.deserialize[DataSource](_, classOf[DataSource]))
-      .map(ds => ds.uri -> ds._key.get)
-      .toMap
-  }
-
-  private def generateNewKeys(uris: Seq[String], uriToExistingKey: Map[String, String]) = {
-    (uris.toSet -- uriToExistingKey.keys)
-        .map(_ -> randomUUID.toString)
-        .toMap
+    Try(attemptSave(entity))
```

Comment:
Are we catching errors from the `attemptSave` method only or the code that is executing asynchronously and is referred by the returned future?
What will happen if the future itself is resolved with an error?

Reply:
fixed